packages/client/src/subscribe/ordering/GapFiller.ts

```diff
@@ -0,0 +1,125 @@
+import { StreamMessage } from '@streamr/protocol'
+import { EthereumAddress, Logger, wait } from '@streamr/utils'
+import sample from 'lodash/sample'
+import { Gap, OrderedMessageChain } from './OrderedMessageChain'
+
+const logger = new Logger(module)
+
+interface GapFillerOptions {
+    chain: OrderedMessageChain
+    resend: (gap: Gap, storageNodeAddress: EthereumAddress, abortSignal: AbortSignal) => AsyncGenerator<StreamMessage>
+    getStorageNodeAddresses: () => Promise<EthereumAddress[]>
+    initialWaitTime: number
+    retryWaitTime: number
+    maxRequestsPerGap: number
+    abortSignal: AbortSignal
+}
+
+const runAbortableTask = async (
+    run: () => Promise<void>
+): Promise<void> => {
+    try {
+        await run()
+    } catch (e) {
+        if (e.code !== 'AbortError') {
+            throw e
+        }
+    }
+}
+
+export class GapFiller {
+
+    private currentTask: { gap: Gap, abortController: AbortController } | undefined = undefined
+    private readonly chain: OrderedMessageChain
+    private readonly resend: (gap: Gap, storageNodeAddress: EthereumAddress, abortSignal: AbortSignal) => AsyncGenerator<StreamMessage>
+    private readonly getStorageNodeAddresses: () => Promise<EthereumAddress[]>
+    private readonly initialWaitTime: number
+    private readonly retryWaitTime: number
+    private readonly maxRequestsPerGap: number
+
+    constructor(opts: GapFillerOptions) {
+        this.chain = opts.chain
+        this.resend = opts.resend
+        this.getStorageNodeAddresses = opts.getStorageNodeAddresses
+        this.initialWaitTime = opts.initialWaitTime
+        this.retryWaitTime = opts.retryWaitTime
+        this.maxRequestsPerGap = opts.maxRequestsPerGap
+        opts.abortSignal.addEventListener('abort', () => {
+            if (this.currentTask !== undefined) {
+                this.currentTask.abortController.abort()
+            }
+        })
+    }
+
+    start(): void {
+        this.chain.on('gapFound', (gap: Gap) => this.onGapFound(gap))
+        this.chain.on('gapResolved', () => this.onGapResolved())
+    }
```

Comment:
Maybe. But as constructors don't usually initialize heavy, it could be a kind of surprise that the whole gap fill process starts just by creating an object instance?